main.go

```diff
@@ -193,6 +195,55 @@ func versionHandler(w http.ResponseWriter, r *http.Request) {
 	w.Write(jsonVersion)
 }
 
+type heartbeat struct {
+	Status bool `json:"status"`
+	Checks struct {
+		CheckAutographHeartbeat bool `json:"check_autograph_heartbeat"`
+	} `json:"checks"`
+	Details string `json:"details"`
+}
+
+// send a GET request to the autograph heartbeat endpoint and
+// evaluate its status code before responding
+func heartbeatHandler(w http.ResponseWriter, r *http.Request) {
+	var st heartbeat
+	// assume the best, change if we encounter errors
+	st.Status = true
+	st.Checks.CheckAutographHeartbeat = true
+
+	u, err := url.Parse(conf.URL)
+	if err != nil {
+		log.Printf("failed to parse conf url %q: %v", conf.URL, err)
+		http.Error(w, "failed to parse conf URL", http.StatusInternalServerError)
+		return
+	}
+	heartbeatURL := fmt.Sprintf("%s://%s/__heartbeat__", u.Scheme, u.Host)
+	resp, err := http.Get(heartbeatURL)
+	if err != nil {
+		st.Checks.CheckAutographHeartbeat = false
+		st.Status = false
+		st.Details = fmt.Sprintf("failed to request autograph heartbeat from %s: %v", heartbeatURL, err)
+		goto done
+	}
+	if resp.StatusCode != http.StatusOK {
+		st.Checks.CheckAutographHeartbeat = false
+		st.Status = false
+		st.Details = fmt.Sprintf("upstream autograph returned heartbeat code %d %s", resp.StatusCode, resp.Status)
+		goto done
+	}
+done:
+	if !st.Status {
+		log.Println(st.Details)
+		w.WriteHeader(http.StatusServiceUnavailable)
+	} else {
+		w.WriteHeader(http.StatusOK)
+	}
+	jsonSt, err := json.Marshal(st)
+	if err != nil {
+		log.Fatal("failed to marshal heartbeat status: %v", err)
```

Comment:
Should write a 5xx here technically. Realistically I can't see us hitting this.

Reply:
I was half considering ignoring `err` given the improbability. The reason for panicking is this is such an improbably error we probably don't want to continue using this instance of the application for real work.